src/shims/sync.rs

```diff
@@ -0,0 +1,438 @@
+    // memory layout: store an i32 in the first four bytes equal to the
+    // corresponding libc mutex kind constant (i.e. PTHREAD_MUTEX_NORMAL)
+
+    fn pthread_mutexattr_init(&mut self, attr_op: OpTy<'tcx, Tag>) -> InterpResult<'tcx, i32> {
+        let this = self.eval_context_mut();
+
+        check_ptr_target_min_size(this, attr_op, 4)?;
+
+        let attr = this.read_scalar(attr_op)?.not_undef()?;
+        if this.is_null(attr)? {
+            return this.eval_libc_i32("EINVAL");
+        }
+
+        let attr_place = this.deref_operand(attr_op)?;
+        let i32_layout = this.layout_of(this.tcx.types.i32)?;
+        let kind_place = attr_place.offset(Size::ZERO, MemPlaceMeta::None, i32_layout, this)?;
+        let default_kind = this.eval_libc("PTHREAD_MUTEX_DEFAULT")?;
+        this.write_scalar(default_kind, kind_place.into())?;
+
+        Ok(0)
+    }
+
+    fn pthread_mutexattr_settype(
+        &mut self,
+        attr_op: OpTy<'tcx, Tag>,
+        kind_op: OpTy<'tcx, Tag>,
+    ) -> InterpResult<'tcx, i32> {
+        let this = self.eval_context_mut();
+
+        check_ptr_target_min_size(this, attr_op, 4)?;
+
+        let attr = this.read_scalar(attr_op)?.not_undef()?;
+        if this.is_null(attr)? {
+            return this.eval_libc_i32("EINVAL");
+        }
+
+        let kind = this.read_scalar(kind_op)?.not_undef()?;
+        if kind == this.eval_libc("PTHREAD_MUTEX_NORMAL")? ||
+                kind == this.eval_libc("PTHREAD_MUTEX_ERRORCHECK")? ||
+                kind == this.eval_libc("PTHREAD_MUTEX_RECURSIVE")? {
+            let attr_place = this.deref_operand(attr_op)?;
+            let i32_layout = this.layout_of(this.tcx.types.i32)?;
+            let kind_place = attr_place.offset(Size::ZERO, MemPlaceMeta::None, i32_layout, this)?;
+            this.write_scalar(kind, kind_place.into())?;
+        } else {
+            let einval = this.eval_libc_i32("EINVAL")?;
+            return Ok(einval);
+        }
+
+        Ok(0)
+    }
+
+    fn pthread_mutexattr_destroy(&mut self, attr_op: OpTy<'tcx, Tag>) -> InterpResult<'tcx, i32> {
+        let this = self.eval_context_mut();
+
+        check_ptr_target_min_size(this, attr_op, 4)?;
+
+        let attr = this.read_scalar(attr_op)?.not_undef()?;
+        if this.is_null(attr)? {
+            return this.eval_libc_i32("EINVAL");
+        }
+
+        let attr_place = this.deref_operand(attr_op)?;
+        let i32_layout = this.layout_of(this.tcx.types.i32)?;
+        let kind_place = attr_place.offset(Size::ZERO, MemPlaceMeta::None, i32_layout, this)?;
+        this.write_scalar(ScalarMaybeUndef::Undef, kind_place.into())?;
+
+        Ok(0)
+    }
+
+    // pthread_mutex_t is between 24 and 48 bytes, depending on the platform
+    // memory layout:
+    // bytes 0-3: reserved for signature on macOS
+    // bytes 4-7: count of how many times this mutex has been locked, as a u32
+    // bytes 12-15: mutex kind, as an i32
```

Comment:
It is not entirely clear to me whether you are describing how this is actually laid out on real platforms, or how you choose to sue this memory here for our emulation. I think it is the latter, but then (a) it should say that (because "depending on the platform memory layout:" makes it sound like this is a real layout), and (b) I don't understand the "signature on macOS" part.